File: Algorithms/src/1722. Minimize Hamming Distance After Swap Operations/Code/solution.java
class Solution {
    int[] parent;
    int[] rank;

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private void union(int a, int b) {
        int pa = find(a);
        int pb = find(b);

        if (pa == pb)
            return;

        if (rank[pa] < rank[pb]) {
            parent[pa] = pb;
        } else if (rank[pb] < rank[pa]) {
            parent[pb] = pa;
        } else {
            parent[pb] = pa;
            rank[pa]++;
        }
    }

    public int minimumHammingDistance(int[] source, int[] target, int[][] allowedSwaps) {
        int n = source.length;

        parent = new int[n];
        rank = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        // Build connected components
        for (int[] swap : allowedSwaps) {
            union(swap[0], swap[1]);
        }

        // Group indices by root parent
        Map<Integer, List<Integer>> groups = new HashMap<>();

        for (int i = 0; i < n; i++) {
            int root = find(i);
            groups.putIfAbsent(root, new ArrayList<>());
            groups.get(root).add(i);
        }

        int answer = 0;

        // Process each component
        for (List<Integer> indices : groups.values()) {
            Map<Integer, Integer> freq = new HashMap<>();

            // Count source values
            for (int idx : indices) {
                freq.put(source[idx], freq.getOrDefault(source[idx], 0) + 1);
            }

            // Match target values
            for (int idx : indices) {
                if (freq.getOrDefault(target[idx], 0) > 0) {
                    freq.put(target[idx], freq.get(target[idx]) - 1);
                } else {
                    answer++;
                }
            }
        }

        return answer;
    }
}

File: Algorithms/src/1722. Minimize Hamming Distance After Swap Operations/Code/solution.go
func minimumHammingDistance(source []int, target []int, allowedSwaps [][]int) int {
	n := len(source)

	parent := make([]int, n)
	rank := make([]int, n)

	for i := 0; i < n; i++ {
		parent[i] = i
	}

	var find func(int) int
	find = func(x int) int {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}

	union := func(a, b int) {
		pa := find(a)
		pb := find(b)

		if pa == pb {
			return
		}

		if rank[pa] < rank[pb] {
			parent[pa] = pb
		} else if rank[pb] < rank[pa] {
			parent[pb] = pa
		} else {
			parent[pb] = pa
			rank[pa]++
		}
	}

	// Build connected components
	for _, swap := range allowedSwaps {
		union(swap[0], swap[1])
	}

	// Group indices by root
	groups := make(map[int][]int)

	for i := 0; i < n; i++ {
		root := find(i)
		groups[root] = append(groups[root], i)
	}

	answer := 0

	// Process each component
	for _, indices := range groups {
		freq := make(map[int]int)

		// Count source values
		for _, idx := range indices {
			freq[source[idx]]++
		}

		// Match target values
		for _, idx := range indices {
			if freq[target[idx]] > 0 {
				freq[target[idx]]--
			} else {
				answer++
			}
		}
	}

	return answer
}